### ISSUE TITLE
SAML authentication appears to not work with Duo IdP

### ISSUE THREAD

**Issue by tbosserman** [kolide/fleet#2004]
### What version of `fleet` are you using (`fleet version --full`)?
fleet - version 2.0.2
  branch: 	changelog
  revision: 	96008f11bb51605cbf22b7bb6f883d262ce13a28
  build date: 	2019-01-18T00:27:22Z
  build user: 	zwass
  go version: 	go1.11.3

### What operating system are you using?
CentOS Linux release 7.6.1810 (Core)

### What did you do?
I authenticated to our IdP. When I use "IdP initiated login" (click on a link in our IdP), it redirects to `/api/v1/kolide/sso/callback`. The Fleet server then just logs this:
```
validation failed: session missing for request
```
If I do "fleet initiated login" (click on the "SIGN ON WITH IDP link on the Fleet login page) it appears to send a malformed / partially formed request to the IdP resulting in this exception on the IdP itself:
```
Exception: Unable to find the current binding.
```

### What did you expect to see?
I expected to be logged in.

### What did you see instead?
In the case of IdP-initiated authentication I just get redirected to the Fleet login page.
In the case of Fleet-initiated authentication I get redirected to a generic error page on the Duo IdP that just says:
```
Oops!
Something went wrong. Please report this error to your IT support team.
```

### Additional Notes
I have tested using Google Chrome and Firefox, both from Linux hosts and from a Macbook.
I also brought up the "Developer Tools" and captured the SAML assertion which was sent to the Fleet server. I can make that available if it would be helpful.

**Comment by zwass** [comment]
Providing the SAML data that is sent between the services would allow me to try to help you debug this. The SAML implementation has worked with a variety of IdPs, so my guess is that you have a misconfiguration.

**Comment by tbosserman** [comment]
I've attempted to attach the SAML assertion sent to Fleet. I'm hoping you can find something that I mis-configured! That's the easiest thing to fix. But I've setup literally dozens of SAML Service Providers. This is the first one that I haven't been able to get to work.
[fleet4.xml.txt](https://github.com/kolide/fleet/files/2953213/fleet4.xml.txt)

**Comment by zwass** [comment]
Is this assertion you provide here the one that Duo sends when you initiate the session from there? Currently Fleet doesn't support provider-initiated sessions, so the error on the Fleet end is to be expected.

Can you provide the assertion that Fleet sends to Duo and/or get more details about the error that occurs on Duo's end?

**Comment by tbosserman** [comment]
I have attached three files:
NOTES-20190303 - Somewhat stream-of-consciousness of an authentication session.
response-raw-20190313 - The raw data that was in the body of the post back to the fleet server.
response-xml-20190313 - That same data turned into "readable" XML (insofar as XML is ever truly readable :-)
[NOTES-20190313.txt](https://github.com/kolide/fleet/files/2964182/NOTES-20190313.txt)
[response-raw-20190313.txt](https://github.com/kolide/fleet/files/2964183/response-raw-20190313.txt)
[response-xml-20190303.txt](https://github.com/kolide/fleet/files/2964184/response-xml-20190303.txt)

Oh, and this is what gets logged by the Fleet server:
```
{"component":"service","err":null,"method":"InitiateSSO","took":"1.561082ms","ts":"2019-03-13T23:42:01.356084587Z"}
{"component":"service","err":"validation failed: session missing for request","method":"CallbackSSO","took":"758.695µs","ts":"2019-03-13T23:42:03.529779853Z"}
{"component":"service","err":null,"method":"SSOSettings","took":"471.618µs","ts":"2019-03-13T23:42:04.301576506Z"}
```

**Comment by zwass** [comment]
@tbosserman Fleet doesn't (currently) support provider-initiated sessions, so that's why you're seeing "validation failed: session missing for request" (per the SAML spec Fleet tries to verify that the request was initiated legitimately by the Fleet instance).

If we can find out more about what Fleet sends to Duo and/or the error that generates on Duo's side then we can try to debug this further.

**Comment by zwass** [comment]
Or am I misunderstanding what you provided here? My interpretation is that `response-xml-20190303.txt` is what Duo sends when you try to initiate the session there?

It looks like `NOTES-20190313.txt` includes some data about the Fleet initiated request so I can dig into that.

**Comment by tbosserman** [comment]
That entire session was from a Service Provider Initiated authentication session. In other words, I clicked on "SIGN ON WITH SQUARE DUO IDP" in Fleet. Everything I wrote about and all those captures are what happened after that. This was *not* an IdP-initiated login.

**Comment by zwass** [comment]
@tbosserman I'm confused because in your initial report you indicate that an error occurs on the IdP when you initiate from Fleet. In these examples it looks like auth is successful on the Duo end and data is sent back to Fleet. Are you still seeing errors on Duo? Do you see an error message on Fleet at this point?

Please feel free to reach out to me on osquery Slack as that may make our back and forth a bit easier.

**Comment by tbosserman** [comment]
Per our slack conversation, attached please find duo-initiated-saml-assertion.txt, which is the SAML assertion which was POSTed to the fleet server during an IdP-initiated authentication attempt.
[duo-initiated-saml-assertion.txt](https://github.com/kolide/fleet/files/2972777/duo-initiated-saml-assertion.txt)

**Comment by zwass** [comment]
@tbosserman and I connected about this on Slack (https://osquery.slack.com/archives/C1XCLA5DZ/p1552678977202500).

I think we have nailed down the issue to Duo's response not containing an `InResponseTo` element in the response. This seems to be required by the spec (because Fleet sends an ID in the request):

<img width="673" alt="Screen Shot 2019-03-15 at 1 50 33 PM" src="https://user-images.githubusercontent.com/575602/54461808-ec4a4680-472a-11e9-9c12-958a360b48bb.png">
<img width="679" alt="Screen Shot 2019-03-15 at 1 49 49 PM" src="https://user-images.githubusercontent.com/575602/54461810-ec4a4680-472a-11e9-8abb-e3cbcdc1abdf.png">
<img width="697" alt="Screen Shot 2019-03-15 at 1 02 37 PM" src="https://user-images.githubusercontent.com/575602/54461811-ec4a4680-472a-11e9-8fbd-cb81105a678a.png">

https://www.oasis-open.org/committees/download.php/35711/sstc-saml-core-errata-2.0-wd-06-diff.pdf
https://www.oasis-open.org/committees/download.php/35389/sstc-saml-profiles-errata-2.0-wd-06-diff.pdf

Now, where to go from here?

Perhaps we can add support to Fleet for IdP-initiated logins, which would provide a workaround for this (possible) bug in the Duo implementation that still conforms to the spec. Or a custom build of Fleet could be created that skips this check as it may not be essential to the security of the protocol (I am not going to weigh in definitively on that point, some discussion available here: https://security.stackexchange.com/questions/42354/do-i-have-to-validate-saml2-inresponseto)

**Comment by tbosserman** [comment]
A followup on this: I worked with the engineers at Duo, and they determined that it actually is a bug in Fleet. They provided me with a recommended patch, and it does, indeed, fix the problem. Here is the patch:

```
diff --git a/server/sso/authorization_request.go b/server/sso/authorization_request.go
index 9c0d0505..9f293090 100644
--- a/server/sso/authorization_request.go
+++ b/server/sso/authorization_request.go
@@ -127,7 +127,7 @@ func deflate(xmlBuffer *bytes.Buffer) (string, error) {
 	if err != nil {
 		return "", errors.Wrap(err, "compressing auth request")
 	}
-	writer.Flush()
+	writer.Close()
 	encbuff := deflated.Bytes()
 	encoded := base64.StdEncoding.EncodeToString(encbuff)
 	return encoded, nil
```
Once I applied that patch and rebuilt Fleet from source, it was able to perform SP-initiated logins using Duo.

**Comment by zwass** [comment]
I merged the above patch in #2025. Did they also make changes on their end to return the `InResponseTo` value? Because without this, Fleet would continue to [throw an error during the SSO callback](https://github.com/kolide/fleet/blob/master/server/service/validation_sessions.go#L13-L17). Or do you have a change removing that check locally in Fleet?

**Comment by tbosserman** [comment]
In an earlier iteration with the Duo engineers, they had told me to make a different change in Fleet's SAML configuration. That change made Duo stop throwing exceptions, but it didn't actually solve the problem. Instead, it *caused* the problem of Fleet throwing the error during the SSO callback. When I applied the above patch to the Fleet code and backed out the previous change I had made to the Fleet SAML configuration, it all started to work properly.
